endpoints/bookstore-grpc/client/src/main/java/com/google/endpoints/examples/bookstore/BookstoreClient.java

```diff
@@ -0,0 +1,235 @@
+      printUsage(options);
+      return;
+    }
+
+    String address = defaultBookstore;
+    String apiKey = null;
+    String authToken = null;
+    String operation = "list";
+
+    if (line.hasOption("bookstore")) {
+      address = line.getOptionValue("bookstore");
+    }
+
+    if (line.hasOption("api_key")) {
+      apiKey = line.getOptionValue("api_key");
+    }
+
+    if (line.hasOption("auth_token")) {
+      authToken = line.getOptionValue("auth_token");
+    }
+
+    if (line.hasOption("operation")) {
+      operation = line.getOptionValue("operation");
+    }
+
+    // Create gRPC stub.
+    BookstoreGrpc.BookstoreBlockingStub bookstore = createBookstoreStub(
+        address, apiKey, authToken);
+
+    if ("list".equals(operation)) {
+      listShelves(bookstore);
+    } else if ("create".equals(operation)) {
+      createShelf(bookstore);
+    } else if ("enumerate".equals(operation)) {
+      enumerate(bookstore);
+    }
+  }
+
+  private static final class Interceptor implements ClientInterceptor {
+    private final String apiKey;
+    private final String authToken;
+
+    private static Metadata.Key<String> apiKeyHeader =
+        Metadata.Key.of("x-api-key", Metadata.ASCII_STRING_MARSHALLER);
+    private static Metadata.Key<String> authorizationHeader =
+        Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
+
+    public Interceptor(String apiKey, String authToken) {
+      this.apiKey = apiKey;
+      this.authToken = authToken;
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+        MethodDescriptor<ReqT,RespT> method, CallOptions callOptions, Channel next) {
+      System.out.println("Intercepted " + method.getFullMethodName());
+      ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+
+      call = new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(call) {
+        @Override
+        public void start(Listener<RespT> responseListener, Metadata headers) {
+          if (apiKey != null && !apiKey.isEmpty()) {
+            System.out.println("Attaching API Key: " + apiKey);
+            headers.put(apiKeyHeader, apiKey);
+          }
+          if (authToken != null && !authToken.isEmpty()) {
+            System.out.println("Attaching auth token");
+            headers.put(authorizationHeader, "Bearer " + authToken);
+          }
+          super.start(responseListener, headers);
+        }
+      };
+      return call;
+    }
+  }
+
+  static BookstoreGrpc.BookstoreBlockingStub createBookstoreStub(
+      String address, String apiKey, String authToken) {
+    Channel channel = ManagedChannelBuilder.forTarget(address)
+        .usePlaintext(true)
+        .build();
+
+    channel = ClientInterceptors.intercept(channel,  new Interceptor(apiKey, authToken));
+
+    return BookstoreGrpc.newBlockingStub(channel);
+  }
+
+  static void listShelves(BookstoreGrpc.BookstoreBlockingStub bookstore) {
```

Comment:
What do you think about moving these three methods to right below the `main` method? This way, reading the program top-down, you can get to the important methods first, before hitting the supporting methods/classes.